### PR TITLE
[Simulator] Initial data logging

### PR DESCRIPTION
**Pull request by microbit-matt-hillsdon**

### CODE REVIEW

File: src/simulator/DataLogging.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    return (
+      <HStack justifyContent="space-between" width="100%">
+        {icon}
+        <Text>{dataLog.data.length} rows</Text>
+        <Button
+          size="sm"
+          leftIcon={<RiDownload2Line />}
+          onClick={handleDownload}
+        >
+          Download
```
Comment (microbit-matt-hillsdon) [inline src/simulator/DataLogging.tsx:49]: This any many other strings need sorting.

Comment (microbit-matt-hillsdon) [reply src/simulator/DataLogging.tsx]: 995966d0a371d14300b047fe376608751250d646

File: src/simulator/Simulator.tsx [hunk 2]
```diff
@@ -32,6 +32,11 @@ const Simulator = ({
   showSimulatorButtonRef,
   minWidth,
 }: SimulatorProps) => {
+  // This needs the domain to be updated before we release.
+  // const url = "https://stage-python-simulator.microbit.org/simulator.html";
+  const url =
+    "https://review-python-simulator.microbit.org/alt-data-logging/simulator.html";
```
[the remark this answers is not in the document]
Comment (microbit-matt-hillsdon) [reply src/simulator/Simulator.tsx:38]: I'll do this as soon as we merge https://github.com/microbit-foundation/micropython-microbit-v2-simulator/pull/41

**Comment by microbit-matt-hillsdon** [comment]
Odd hang in CI for a comment only change. Re-running.

File: lang/ui.en.json [hunk 3]
```diff
@@ -247,6 +247,10 @@
     "defaultMessage": "Don't show this again",
     "description": "Text to never show a dialog again"
   },
+  "download-action": {
+    "defaultMessage": "Download",
+    "description": "Download action label"
+  },
```
Comment (microbit-robert) [inline lang/ui.en.json:253]: Not used.

Comment (microbit-matt-hillsdon) [reply lang/ui.en.json:253]: Thanks!

File: src/simulator/Simulator.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (microbit-matt-hillsdon) [reply src/simulator/Simulator.tsx]: 072dc9c608c162e575c3b00fdc428d99b13b9abc

File: lang/ui.en.json (hunk 3 above)
Comment (microbit-matt-hillsdon) [reply lang/ui.en.json]: aefa21dd3572f9fa2fa605b70ff5831e959acdb7

File: src/simulator/data-logging-hooks.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+export interface DataLog {
+  headings: string[];
+  data: DataLogRow[];
+}
+
+export interface DataLogRow {
+  isHeading?: boolean;
+  data: string[];
+}
```
Comment (microbit-robert) [inline src/simulator/data-logging-hooks.tsx:14]: These are also defined identically in and exported from `simulator.ts`.

Comment (microbit-matt-hillsdon) [reply src/simulator/data-logging-hooks.tsx:14]: Oops, I intended to move it there when I moved the log management there for performance reasons.

Comment (microbit-matt-hillsdon) [reply src/simulator/data-logging-hooks.tsx]: 3360b3f57759fb27bb5224cb61db201fd37e658a

**Comment by microbit-robert** [comment]
I have issues with the automatic scroll behaviour. It seems to work once and then never catches up?

**Comment by microbit-matt-hillsdon** [comment]
> I have issues with the automatic scroll behaviour. It seems to work once and then never catches up?

Pushed a fix. It is still possible to break if messages come faster than the smooth scroll speed, but not addressing that for now.

**Review by microbit-robert — APPROVED** [review]
LGTM.

**Pull request merged**